js/ui/light.js

```diff
@@ -0,0 +1,256 @@
+var interpolate = require('../util/interpolate');
+var parseColor = require('../style/parse_color');
+
+/**
+ * Controls the lighting used to light extruded features.
+ *
+ * @typedef {Object} LightOptions
+ * @property {string} anchor Whether the light direction should be oriented based on the map or viewport. Options are `'map'`, `'viewport'`.
+ * @property {Color} color The color to tint the extrusion lighting.
+ * @property {Array<number>} direction The direction of the light source, in [r radial coordinate, θ azimuthal angle, φ polar angle], where r indicates the distance from the center of an object to its light, θ indicates the position of the light relative to 0° (like a clock), and φ indicates the height of the light source (from 0°, directly above, to 180°, directly below).
+ * @property {number} intensity The intensity with which to light extruded features.
+ * @property {number} duration The lighting animation's duration, measured in milliseconds.
+ * @property {Function} easing The lighting animation's easing function.
+ * @property {boolean} animate If `false`, no lighting animation will occur.
+ */
+
+var Lighting = module.exports = function() {};
+
+util.extend(Lighting.prototype, /** @lends Map.prototype */{
+    activeLightTransitions: {},
+
+    /**
+     * Get the full lighting object.
+     *
+     * @returns {Object} light Light object as denoted in the style spec.
+     */
+    getLight: function() {
+        return {
+            anchor: this.getLightProperty('anchor'),
+            color: this.getLightProperty('color'),
+            direction: this.getLightProperty('direction'),
+            intensity: this.getLightProperty('intensity')
+        };
+    },
+
+    /**
+     * Get a specific light property.
+     *
+     * @param {String} property Light property. One of `anchor`, `color`, `direction`, `intensity`.
+     * @returns {Value} value Value for specified property. Type per property is denoted in the style spec.
+     */
+    getLightProperty: function(property) {
+        if (!(property in this.painter.light)) throw new Error('Unrecognized light property: ' + property);
+
+        if (property === 'direction') {
+            var dir = this.painter.light.direction;
+            return [dir.r, dir.a, dir.p];
+        }
+
+        if (property === 'color') {
+            return this.painter.light.rawColor;
+        }
+
+        return this.painter.light[property];
+    },
+
+    /**
+     * Changes any combination of light values, with an animated transition
+     * between old and new values (with the exception of `anchor`, which does
+     * not animate). The map will retain its current values for any
+     * details not specified in `options`.
+     *
+     * @param {LightOptions} options Options describing the target light properties of the transition.
+     * @param {Object} [eventData] Data to propagate to any event listeners.
+     * @fires lightstart
+     * @fires lightend
```

Comment:
I like the idea of an events that  mark the beginning and end of a style property transition. I'm not sure we need such an event specifically for `light` properties, however. What do yout think about a follow-up PR that adds `transitionstart` and `transitionend` events?